File: labs/M17-output-guardrails-hitl/starter/output_validator.py
"""
M17: Output Validator — Starter
Validates agent outputs for structure, hallucination markers, and PII leakage.
"""
import re
import json


# ── Hallucination Markers ───────────────────────────────────
# Phrases that signal the agent is guessing rather than answering from data
HALLUCINATION_MARKERS = [
    # TODO 1: Add at least 6 low-confidence phrases that indicate hallucination.
    # Examples: "I think", "probably", "I'm not sure", "it seems like",
    #           "I believe", "might be", "not entirely certain"
    # Format: list of lowercase strings
]


# ── PII Patterns for Output Scanning ───────────────────────
PII_PATTERNS = [
    ("ssn", r"\b\d{3}-\d{2}-\d{4}\b", "[SSN REDACTED]"),
    ("credit_card", r"\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b", "[CC REDACTED]"),
    ("email", r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b", "[EMAIL REDACTED]"),
    ("phone", r"\b(?:\+1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b", "[PHONE REDACTED]"),
]


def check_json_structure(output: dict, required_fields: list[str]) -> dict:
    """
    Validate that the output dict contains all required fields.

    Returns:
        {
            "valid": bool,
            "missing_fields": list[str],
            "extra_fields": list[str]
        }
    """
    # TODO 2: Check which required_fields are missing from output keys.
    # Also identify any unexpected fields not in the required list.
    # Return the result dict with valid=True only if no fields are missing.
    missing = []
    extra = []

    return {
        "valid": len(missing) == 0,
        "missing_fields": missing,
        "extra_fields": extra,
    }


def check_hallucination_markers(text: str) -> dict:
    """
    Scan text for low-confidence phrases indicating potential hallucination.

    Returns:
        {
            "has_markers": bool,
            "markers_found": list[str],
            "confidence_penalty": float  # 0.1 per marker found, max 0.5
        }
    """
    # TODO 3: Check text (lowercased) for each marker in HALLUCINATION_MARKERS.
    # Collect all found markers.
    # Calculate confidence_penalty: 0.1 per marker, capped at 0.5.
    markers_found = []
    penalty = 0.0

    return {
        "has_markers": len(markers_found) > 0,
        "markers_found": markers_found,
        "confidence_penalty": penalty,
    }


def check_pii_in_output(text: str) -> dict:
    """
    Ensure agent outputs don't leak PII data.

    Returns:
        {
            "has_pii": bool,
            "pii_types": list[str],
            "redacted_text": str
        }
    """
    # TODO 4: Loop through PII_PATTERNS. Use re.findall to detect matches.
    # Collect the PII type names found.
    # Replace all PII matches in text with the redaction string.
    pii_types = []
    redacted = text

    return {
        "has_pii": len(pii_types) > 0,
        "pii_types": pii_types,
        "redacted_text": redacted,
    }


def validate_output(output: dict, expected_fields: list[str]) -> dict:
    """
    Run all output validation checks.

    Returns:
        {
            "valid": bool,
            "checks": {
                "structure": {...},
                "hallucination": {...},
                "pii": {...}
            },
            "output": dict  # original or redacted output
        }
    """
    # TODO 5: Run check_json_structure with output and expected_fields.
    # TODO 6: If output has a "response" key (string), run check_hallucination_markers
    #         and check_pii_in_output on it. If PII found, replace the response with
    #         the redacted version.
    # Combine all check results. valid=True only if structure is valid AND no PII found.

    structure_check = {"valid": True, "missing_fields": [], "extra_fields": []}
    hallucination_check = {"has_markers": False, "markers_found": [], "confidence_penalty": 0.0}
    pii_check = {"has_pii": False, "pii_types": [], "redacted_text": ""}

    is_valid = structure_check["valid"] and not pii_check["has_pii"]

    return {
        "valid": is_valid,
        "checks": {
            "structure": structure_check,
            "hallucination": hallucination_check,
            "pii": pii_check,
        },
        "output": output,
    }


# ── Self-Test ───────────────────────────────────────────────
if __name__ == "__main__":
    print("=" * 60)
    print("M17 Output Validator — Self-Test")
    print("=" * 60)

    # Test 1: Valid output
    good_output = {
        "entity": "Acme Corp",
        "filing_number": "UCC-2024-CA-0001234",
        "response": "Acme Corp has 3 active UCC filings in California.",
        "confidence": 0.95,
    }
    result = validate_output(good_output, ["entity", "filing_number", "response", "confidence"])
    print(f"\nTest 1 — Valid output: {'PASS' if result['valid'] else 'FAIL'}")
    print(f"  Structure valid: {result['checks']['structure']['valid']}")
    print(f"  Hallucination markers: {result['checks']['hallucination']['has_markers']}")
    print(f"  PII detected: {result['checks']['pii']['has_pii']}")

    # Test 2: Missing fields
    bad_structure = {"entity": "Acme Corp"}
    result = validate_output(bad_structure, ["entity", "filing_number", "response", "confidence"])
    print(f"\nTest 2 — Missing fields: {'PASS' if not result['valid'] else 'FAIL'}")
    print(f"  Missing: {result['checks']['structure']['missing_fields']}")

    # Test 3: Hallucination markers
    hedging_output = {
        "entity": "Acme Corp",
        "filing_number": "UCC-2024-CA-0001234",
        "response": "I think Acme Corp probably has some filings, but I'm not sure about the details.",
        "confidence": 0.75,
    }
    result = validate_output(hedging_output, ["entity", "filing_number", "response", "confidence"])
    print(f"\nTest 3 — Hallucination markers:")
    print(f"  Markers found: {result['checks']['hallucination']['markers_found']}")
    print(f"  Confidence penalty: {result['checks']['hallucination']['confidence_penalty']}")

    # Test 4: PII in output
    pii_output = {
        "entity": "John Smith",
        "filing_number": "UCC-2024-NY-0005678",
        "response": "Contact John Smith at 555-123-4567 or john@example.com. SSN: 123-45-6789.",
        "confidence": 0.85,
    }
    result = validate_output(pii_output, ["entity", "filing_number", "response", "confidence"])
    print(f"\nTest 4 — PII in output: {'PASS' if not result['valid'] else 'FAIL'}")
    print(f"  PII types: {result['checks']['pii']['pii_types']}")
    print(f"  Redacted: {result['output'].get('response', 'N/A')}")

    print("\n" + "=" * 60)
    print("Self-test complete. Fill in TODOs to make all tests pass.")
    print("=" * 60)

File: labs/M17-output-guardrails-hitl/starter/output_validator.js
/**
 * M17: Output Validator — Starter
 * Validates agent outputs for structure, hallucination markers, and PII leakage.
 */

// ── Hallucination Markers ───────────────────────────────────
// Phrases that signal the agent is guessing rather than answering from data
const HALLUCINATION_MARKERS = [
  // TODO 1: Add at least 6 low-confidence phrases that indicate hallucination.
  // Examples: "i think", "probably", "i'm not sure", "it seems like",
  //           "i believe", "might be", "not entirely certain"
];

// ── PII Patterns for Output Scanning ───────────────────────
const PII_PATTERNS = [
  { name: "ssn", pattern: /\b\d{3}-\d{2}-\d{4}\b/g, replacement: "[SSN REDACTED]" },
  { name: "credit_card", pattern: /\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b/g, replacement: "[CC REDACTED]" },
  { name: "email", pattern: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g, replacement: "[EMAIL REDACTED]" },
  { name: "phone", pattern: /\b(?:\+1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/g, replacement: "[PHONE REDACTED]" },
];

/**
 * Validate that the output object contains all required fields.
 * @param {Object} output - Agent output object.
 * @param {string[]} requiredFields - List of required field names.
 * @returns {{ valid: boolean, missingFields: string[], extraFields: string[] }}
 */
export function checkJsonStructure(output, requiredFields) {
  // TODO 2: Check which requiredFields are missing from Object.keys(output).
  // Also identify any keys in output not in requiredFields.
  // Return valid=true only if no fields are missing.
  const missing = [];
  const extra = [];

  return {
    valid: missing.length === 0,
    missingFields: missing,
    extraFields: extra,
  };
}

/**
 * Scan text for low-confidence phrases indicating potential hallucination.
 * @param {string} text - Text to scan.
 * @returns {{ hasMarkers: boolean, markersFound: string[], confidencePenalty: number }}
 */
export function checkHallucinationMarkers(text) {
  // TODO 3: Check text (lowercased) for each marker in HALLUCINATION_MARKERS.
  // Collect all found markers.
  // Calculate confidencePenalty: 0.1 per marker, capped at 0.5.
  const markersFound = [];
  const penalty = 0.0;

  return {
    hasMarkers: markersFound.length > 0,
    markersFound,
    confidencePenalty: penalty,
  };
}

/**
 * Ensure agent outputs don't leak PII data.
 * @param {string} text - Text to scan.
 * @returns {{ hasPii: boolean, piiTypes: string[], redactedText: string }}
 */
export function checkPiiInOutput(text) {
  // TODO 4: Loop through PII_PATTERNS. Use pattern.test() to detect matches.
  // Collect the PII type names found.
  // Replace all PII matches in text with the replacement string.
  // IMPORTANT: Reset regex lastIndex before each test since /g flag is used.
  const piiTypes = [];
  let redacted = text;

  return {
    hasPii: piiTypes.length > 0,
    piiTypes,
    redactedText: redacted,
  };
}

/**
 * Run all output validation checks.
 * @param {Object} output - Agent output object.
 * @param {string[]} expectedFields - Expected field names.
 * @returns {{ valid: boolean, checks: Object, output: Object }}
 */
export function validateOutput(output, expectedFields) {
  // TODO 5: Run checkJsonStructure with output and expectedFields.
  // TODO 6: If output has a "response" key (string), run checkHallucinationMarkers
  //         and checkPiiInOutput on it. If PII found, replace the response with
  //         the redacted version in a copy of the output.
  // Combine all check results. valid=true only if structure is valid AND no PII found.

  const structureCheck = { valid: true, missingFields: [], extraFields: [] };
  const hallucinationCheck = { hasMarkers: false, markersFound: [], confidencePenalty: 0.0 };
  const piiCheck = { hasPii: false, piiTypes: [], redactedText: "" };

  const isValid = structureCheck.valid && !piiCheck.hasPii;

  return {
    valid: isValid,
    checks: {
      structure: structureCheck,
      hallucination: hallucinationCheck,
      pii: piiCheck,
    },
    output,
  };
}

// ── Self-Test ───────────────────────────────────────────────
const isMain = process.argv[1] && (
  process.argv[1].endsWith("output_validator.js") ||
  process.argv[1].endsWith("output_validator.mjs")
);

if (isMain) {
  console.log("=".repeat(60));
  console.log("M17 Output Validator — Self-Test");
  console.log("=".repeat(60));

  // Test 1: Valid output
  const goodOutput = {
    entity: "Acme Corp",
    filing_number: "UCC-2024-CA-0001234",
    response: "Acme Corp has 3 active UCC filings in California.",
    confidence: 0.95,
  };
  let result = validateOutput(goodOutput, ["entity", "filing_number", "response", "confidence"]);
  console.log(`\nTest 1 — Valid output: ${result.valid ? "PASS" : "FAIL"}`);
  console.log(`  Structure valid: ${result.checks.structure.valid}`);
  console.log(`  Hallucination markers: ${result.checks.hallucination.hasMarkers}`);
  console.log(`  PII detected: ${result.checks.pii.hasPii}`);

  // Test 2: Missing fields
  const badStructure = { entity: "Acme Corp" };
  result = validateOutput(badStructure, ["entity", "filing_number", "response", "confidence"]);
  console.log(`\nTest 2 — Missing fields: ${!result.valid ? "PASS" : "FAIL"}`);
  console.log(`  Missing: ${JSON.stringify(result.checks.structure.missingFields)}`);

  // Test 3: Hallucination markers
  const hedgingOutput = {
    entity: "Acme Corp",
    filing_number: "UCC-2024-CA-0001234",
    response: "I think Acme Corp probably has some filings, but I'm not sure about the details.",
    confidence: 0.75,
  };
  result = validateOutput(hedgingOutput, ["entity", "filing_number", "response", "confidence"]);
  console.log(`\nTest 3 — Hallucination markers:`);
  console.log(`  Markers found: ${JSON.stringify(result.checks.hallucination.markersFound)}`);
  console.log(`  Confidence penalty: ${result.checks.hallucination.confidencePenalty}`);

  // Test 4: PII in output
  const piiOutput = {
    entity: "John Smith",
    filing_number: "UCC-2024-NY-0005678",
    response: "Contact John Smith at 555-123-4567 or john@example.com. SSN: 123-45-6789.",
    confidence: 0.85,
  };
  result = validateOutput(piiOutput, ["entity", "filing_number", "response", "confidence"]);
  console.log(`\nTest 4 — PII in output: ${!result.valid ? "PASS" : "FAIL"}`);
  console.log(`  PII types: ${JSON.stringify(result.checks.pii.piiTypes)}`);
  console.log(`  Redacted: ${result.output.response || "N/A"}`);

  console.log("\n" + "=".repeat(60));
  console.log("Self-test complete. Fill in TODOs to make all tests pass.");
  console.log("=".repeat(60));
}
